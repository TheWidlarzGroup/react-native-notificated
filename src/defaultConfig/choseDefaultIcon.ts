import type { ImageSourcePropType } from 'react-native'
import type {IconsLinksTypes, IconVisualStyle, NotificationVariants} from "./types";

export const chooseDefaultIcon = (
  notificationType: NotificationVariants,
  darkMode: boolean,
  defaultIconType?: IconVisualStyle
): ImageSourcePropType | undefined => {
  const iconLinks: IconsLinksTypes = {
    success: {
      color: require(`../assets/images/success-color.png`),
      white: require(`../assets/images/success-white.png`),
      black: require(`../assets/images/success-black.png`),
    },
    error: {
      color: require(`../assets/images/error-color.png`),
      white: require(`../assets/images/error-white.png`),
      black: require(`../assets/images/error-black.png`),
    },
    warning: {
      color: require(`../assets/images/warning-color.png`),
      white: require(`../assets/images/warning-white.png`),
      black: require(`../assets/images/warning-black.png`),
    },
    info: {
      color: require(`../assets/images/info-color.png`),
      white: require(`../assets/images/info-white.png`),
      black: require(`../assets/images/info-black.png`),
    },
  }

  const renderIcon = (type: NotificationVariants): ImageSourcePropType | undefined => {
    switch (defaultIconType) {
      case 'color':
        return iconLinks[type].color
      case 'monochromatic':
        return darkMode ? iconLinks[type].white : iconLinks[type].black
      case 'no-icon':
        return undefined
      default:
        return iconLinks[type].color
    }
  }

  switch (notificationType) {
    case 'success':
      return renderIcon('success')
    case 'error':
      return renderIcon('error')
    case 'warning':
      return renderIcon('warning')
    case 'info':
      return renderIcon('info')
  }
}
