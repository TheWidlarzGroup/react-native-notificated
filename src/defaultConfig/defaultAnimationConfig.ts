import { Easing, interpolate } from 'react-native-reanimated'
import { generateAnimationConfig } from '../core/utils/generateAnimationConfig'

export const ZoomInDownZoomOutUp = generateAnimationConfig({
  animationConfigIn: {
    type: 'timing',
    config: {
      duration: 400,
      easing: Easing.inOut(Easing.sin),
    },
  },
  transitionInStyles: (progress) => {
    'worklet'

    const scale = interpolate(progress.value, [0, 1], [0.8, 1])
    const translateY = interpolate(progress.value, [0, 1], [-100, 0])

    return {
      opacity: progress.value,
      transform: [{ scale }, { translateY }],
    }
  },
})

export const ZoomInDownZoomOutDown = generateAnimationConfig({
  animationConfigIn: {
    type: 'spring',
    config: {
      damping: 4,
      mass: 0.8,
    },
  },
  transitionInStyles: (progress) => {
    'worklet'

    const scale = interpolate(progress.value, [0, 1], [0.8, 1])
    const translateY = interpolate(progress.value, [0, 1], [-100, 0])

    return {
      opacity: progress.value,
      transform: [{ scale }, { translateY }],
    }
  },
  transitionOutStyles: (progress) => {
    'worklet'

    const scale = interpolate(progress.value, [0, 1], [0.8, 1])
    const translateY = interpolate(progress.value, [0, 1], [100, 0])

    return {
      opacity: progress.value,
      transform: [{ scale }, { translateY }],
    }
  },
})

export const RotateInRotateOut = generateAnimationConfig({
  animationConfigIn: {
    type: 'timing',
    config: {
      duration: 700,
      easing: Easing.out(Easing.exp),
    },
  },
  transitionInStyles: (progress) => {
    'worklet'

    const rotate = interpolate(progress.value, [0, 1], [-360, 0])

    return {
      transform: [{ rotate: `${rotate}deg` }, { scale: progress.value }],
      opacity: progress.value,
    }
  },
})

export const RotateZIn = generateAnimationConfig({
  animationConfigIn: {
    type: 'timing',
    config: {
      duration: 2000,
      easing: Easing.out(Easing.exp),
    },
  },
  transitionInStyles: (progress) => {
    'worklet'

    const rotateZ = interpolate(progress.value, [0, 1], [-140, 0])
    const rotateX = interpolate(progress.value, [0, 1], [200, 20])
    const rotateY = interpolate(progress.value, [0, 1], [-40, 5])

    return {
      transform: [
        { rotateZ: `${rotateZ}deg` },
        { rotateX: `${rotateX}deg` },
        { rotateY: `${rotateY}deg` },
        { scale: progress.value },
        { perspective: 1500 },
      ],
      opacity: progress.value,
    }
  },
})

export const SlideInLeftSlideOutRight = generateAnimationConfig({
  animationConfigIn: {
    type: 'timing',
    config: {
      duration: 400,
      easing: Easing.inOut(Easing.sin),
    },
  },
  transitionInStyles: (progress) => {
    'worklet'

    const translateX = interpolate(progress.value, [0, 1], [-100, 0])

    return {
      opacity: progress.value,
      transform: [{ translateX }],
    }
  },
  transitionOutStyles: (progress) => {
    'worklet'

    const translateX = interpolate(progress.value, [0, 1], [100, 0])

    return {
      opacity: progress.value,
      transform: [{ translateX }],
    }
  },
})

export const CrazyAnimationConfig = generateAnimationConfig({
  animationConfigIn: {
    type: 'spring',
    config: { damping: 10, velocity: 20, stiffness: 80, mass: 1.2 },
  },
  animationConfigOut: {
    type: 'timing',
    config: { duration: 200 },
  },
  transitionInStyles: (progress) => {
    'worklet'

    const interpolatedTrans = interpolate(progress.value, [0, 1], [0, 100])

    return {
      transform: [{ translateY: 500 }, { translateX: interpolatedTrans }],
      opacity: progress.value,
    }
  },
  transitionOutStyles: (progress) => {
    'worklet'

    const interpolatedTrans = interpolate(progress.value, [0, 1], [0, 500])

    return {
      transform: [{ translateY: interpolatedTrans }],
      opacity: progress.value,
    }
  },
})

export const VeryCustomTransition = generateAnimationConfig({
  animationConfigIn: {
    type: 'spring',
    config: { damping: 10, velocity: 20, stiffness: 80, mass: 1.2 },
  },
  animationConfigOut: {
    type: 'timing',
    config: { duration: 200 },
  },
  transitionInStyles: (progress) => {
    'worklet'

    const translateY = interpolate(progress.value, [0, 1], [0, 100])

    return {
      transform: [{ translateY }],
      opacity: progress.value,
    }
  },
  transitionOutStyles: (progress) => {
    'worklet'

    const translateY = interpolate(progress.value, [0, 1], [500, 100])

    return {
      transform: [{ translateY }],
      opacity: progress.value,
    }
  },
})

export const DiagonalSlideInLeftSlideOutRight = generateAnimationConfig({
  animationConfigIn: {
    type: 'timing',
    config: {
      duration: 700,
    },
  },
  transitionInStyles: (progress) => {
    'worklet'

    const translateX = interpolate(progress.value, [0, 1], [-100, 0])
    const translateY = interpolate(progress.value, [0, 1], [-100, 0])
    return {
      transform: [{ translateX }, { translateY }],
      opacity: progress.value,
    }
  },
  transitionOutStyles: (progress) => {
    'worklet'

    const translateX = interpolate(progress.value, [0, 1], [100, 0])
    const translateY = interpolate(progress.value, [0, 1], [-100, 0])
    return {
      transform: [{ translateX }, { translateY }],
      opacity: progress.value,
    }
  },
})
