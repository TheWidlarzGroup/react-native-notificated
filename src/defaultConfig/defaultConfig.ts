import { SuccessNotification } from './components/success'
import { WarningNotification } from './components/warning'
import { ErrorNotification } from './components/error'
import { InfoNotification } from './components/info'
import type { DefaultLayoutConfig, NotificationsConfig, Variant } from '../types'

export type _DefaultVariants = {
  success: Variant<typeof SuccessNotification>
  error: Variant<typeof ErrorNotification>
  warning: Variant<typeof WarningNotification>
  info: Variant<typeof InfoNotification>
}

export const defaultVariants: _DefaultVariants = {
  success: {
    component: SuccessNotification,
  },
  warning: {
    component: WarningNotification,
  },
  error: {
    component: ErrorNotification,
  },
  info: {
    component: InfoNotification,
  },
} as const

export const InAppNotificationsConfig: NotificationsConfig<_DefaultVariants> &
  Omit<DefaultLayoutConfig, 'variants'> = {
  defaultNotificationTime: 3000,
  defaultNotificationTimeLong: 5000,
  notificationMsgLengthTimerThreshold: 100,

  variants: defaultVariants,
} as const

export type DefaultKeys = keyof typeof defaultVariants
