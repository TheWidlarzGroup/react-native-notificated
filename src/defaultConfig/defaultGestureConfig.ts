import type { GestureConfig } from '../types/gestures'

export const AndroidGestureConfig: GestureConfig = {
  direction: 'x',
  activationDistances: 50,
  activationVelocities: 2000,
}

export const IosGestureConfig: GestureConfig = {
  direction: 'y',
  activationDistances: 50,
  activationVelocities: 300,
}

export const FrozenGestureConfig: GestureConfig = {
  direction: 'none',
}

export const FullGestureConfig: GestureConfig = {
  direction: 'full',
  x: {
    activationDistances: 100,
    activationVelocities: 2000,
  },
  y: {
    activationDistances: 100,
    activationVelocities: 1000,
  },
}
