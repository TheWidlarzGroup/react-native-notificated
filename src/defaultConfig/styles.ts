import { StyleSheet } from 'react-native'
import { themeBase } from './components/theme'

export const styles = StyleSheet.create({
  icon: {
    height: 35,
    width: 35,
  },
  textWrapper: {
    flex: 1,
    paddingLeft: themeBase.spacing.s,
    paddingRight: themeBase.spacing.l,
  },
  content: {
    paddingVertical: themeBase.spacing.xl,
    paddingLeft: themeBase.spacing.s,
    paddingRight: themeBase.spacing.l,
    display: 'flex',
    justifyContent: 'flex-start',
    alignItems: 'center',
    flex: 1,
    flexDirection: 'row',
  },
  undo: {
    fontSize: themeBase.fontSize.messageFontSize,
    fontWeight: 'bold',
    color: themeBase.color.info,
    textDecorationLine: 'underline',
    paddingTop: themeBase.spacing.xs,
  },
})
