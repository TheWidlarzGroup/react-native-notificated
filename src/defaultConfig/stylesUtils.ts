import type { TextStyle, ViewStyle } from 'react-native'
import { Platform } from 'react-native'
import { themeBase } from './components/theme'
import type { MergedNotificationStyleConfig, NotificationVariants, Theme } from './types'

export const getTitleStyle = (styles: MergedNotificationStyleConfig): Partial<TextStyle> => ({
  color: styles.titleColor ? styles.titleColor : themeBase.fontColor[styles.theme],
  fontSize: styles.titleSize ? styles.titleSize : themeBase.fontSize.headerFontSize,
  paddingBottom: themeBase.spacing.xs,
  flex: 1,
})

export const getDescriptionStyle = (styles: MergedNotificationStyleConfig): Partial<TextStyle> => ({
  color: styles.descriptionColor ? styles.descriptionColor : themeBase.fontColor[styles.theme],
  fontSize: styles.descriptionSize ? styles.descriptionSize : themeBase.fontSize.messageFontSize,
})

export const constShadow = (theme: Theme, borderRadius?: number): Partial<ViewStyle> => {
  const crossPlatformStyle = Platform.select({
    ios: {
      shadowColor: themeBase.color.shadow,
      shadowOffset: { width: 1, height: 2 },
      shadowOpacity: 1,
      shadowRadius: 7,
    },
    android: {
      elevation: 5,
      borderBottomWidth: 0,
      borderRadius: borderRadius,
    },
  })

  return theme === 'regular' && crossPlatformStyle ? crossPlatformStyle : {}
}

export const getContainerStyles = (styles: MergedNotificationStyleConfig): Partial<ViewStyle> => {
  const defaultBackgroundColor = styles.theme
    ? themeBase.bgColor[styles.theme]
    : themeBase.bgColor.regular

  return {
    ...styles,
    display: 'flex',
    flexDirection: 'row',
    justifyContent: 'flex-start',
    alignItems: 'center',
    overflow: 'hidden',
    borderRadius: styles.borderRadius ?? themeBase.borderRadius.regular,
    borderWidth: styles.borderType === 'border' ? styles.borderWidth : 0,
    borderColor: styles.accentColor,
    backgroundColor: styles.bgColor ?? defaultBackgroundColor,
  }
}

export const getLeftAccentStyle = (accentColor: string | undefined) => {
  return {
    flex: 0.045,
    height: '100%',
    backgroundColor: accentColor ?? '',
  }
}

export const chooseDefaultAccentColor = (notificationType: NotificationVariants) => {
  switch (notificationType) {
    case 'success':
      return themeBase.color.success
    case 'error':
      return themeBase.color.error
    case 'warning':
      return themeBase.color.warning
    case 'info':
      return themeBase.color.info
  }
}
