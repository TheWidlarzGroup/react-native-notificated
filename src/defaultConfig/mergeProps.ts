import type {
  MergedNotificationStyleConfig,
  NotificationProps,

} from '../types'
import type { ImageSourcePropType } from 'react-native'
import { themeBase } from './components/theme'
import { chooseDefaultAccentColor } from './stylesUtils'
import { chooseDefaultIcon } from './choseDefaultIcon'
import type {NotificationStyleConfig, NotificationVariants} from "./types";

export const mergeProps = (
  props: NotificationProps,
  notificationType: NotificationVariants,
  darkMode: boolean,
  defaultGlobalConfig?: NotificationStyleConfig,
  defaultNotificationTypeConfig?: NotificationStyleConfig
): NotificationProps & MergedNotificationStyleConfig => {
  const customIconSource: ImageSourcePropType | undefined =
    props.leftIconSource ??
    defaultNotificationTypeConfig?.leftIconSource ??
    defaultGlobalConfig?.leftIconSource

  return {
    title: props.title ?? '',
    description: props.description ?? '',
    theme: darkMode ? 'dark' : 'regular',
    titleSize:
      props.titleSize ?? defaultNotificationTypeConfig?.titleSize ?? defaultGlobalConfig?.titleSize,
    titleColor:
      props.titleColor ??
      defaultNotificationTypeConfig?.titleColor ??
      defaultGlobalConfig?.titleColor,
    descriptionColor:
      props.descriptionColor ??
      defaultNotificationTypeConfig?.descriptionColor ??
      defaultGlobalConfig?.descriptionColor,
    descriptionSize:
      props.descriptionSize ??
      defaultNotificationTypeConfig?.descriptionSize ??
      defaultGlobalConfig?.descriptionSize,
    bgColor:
      props.bgColor ?? defaultNotificationTypeConfig?.bgColor ?? defaultGlobalConfig?.bgColor,
    borderRadius:
      props.borderRadius ??
      defaultNotificationTypeConfig?.borderRadius ??
      (defaultGlobalConfig?.borderRadius || themeBase.borderRadius.regular),
    accentColor:
      props.accentColor ??
      defaultNotificationTypeConfig?.accentColor ??
      (defaultGlobalConfig?.accentColor || chooseDefaultAccentColor(notificationType)),
    borderWidth:
      props.borderWidth ??
      defaultNotificationTypeConfig?.borderWidth ??
      defaultGlobalConfig?.borderWidth ??
      1,
    multiline:
      props.multiline ?? defaultNotificationTypeConfig?.multiline ?? defaultGlobalConfig?.multiline,
    defaultIconType:
      props.defaultIconType ??
      defaultNotificationTypeConfig?.defaultIconType ??
      defaultGlobalConfig?.defaultIconType,
    leftIconSource:
      customIconSource ??
      chooseDefaultIcon(
        notificationType,
        darkMode,
        props.defaultIconType ??
          defaultNotificationTypeConfig?.defaultIconType ??
          defaultGlobalConfig?.defaultIconType
      ),
    borderType:
      props.borderType ??
      defaultNotificationTypeConfig?.borderType ??
      defaultGlobalConfig?.borderType ??
      'border',
    notificationType: notificationType,
    onPress: props.onPress ?? undefined,
  }
}
