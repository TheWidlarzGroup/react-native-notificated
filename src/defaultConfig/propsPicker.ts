import { themeBase } from './components/theme'
import type { PropsConfig, DefaultStylesConfig } from '../types'
import { chooseDefaultIcon } from './choseDefaultIcon'
import type { NotificationVariants } from '../types'
import { chooseAccentColor } from './stylesUtils'

export const mergeProps = (
  props: Partial<PropsConfig>,
  notificationType: NotificationVariants,
  darkMode: boolean,
  defaultGlobalConfig?: DefaultStylesConfig,
  defaultNotificationTypeConfig?: DefaultStylesConfig
): PropsConfig => {
  const customIconSource =
    props.leftIconSource ??
    defaultNotificationTypeConfig?.leftIconSource ??
    defaultGlobalConfig?.leftIconSource

  return {
    title: props.title ?? '',
    description: props.description ?? 'Description',
    theme: darkMode ? 'dark' : 'regular',
    titleSize:
      props.titleSize ?? defaultNotificationTypeConfig?.titleSize ?? defaultGlobalConfig?.titleSize,
    titleColor:
      props.titleColor ??
      defaultNotificationTypeConfig?.titleColor ??
      defaultGlobalConfig?.titleColor,
    descriptionColor:
      props.descriptionColor ??
      defaultNotificationTypeConfig?.descriptionColor ??
      defaultGlobalConfig?.descriptionColor,
    descriptionSize:
      props.descriptionSize ??
      defaultNotificationTypeConfig?.descriptionSize ??
      defaultGlobalConfig?.descriptionSize,
    bgColor:
      props.bgColor ?? defaultNotificationTypeConfig?.bgColor ?? defaultGlobalConfig?.bgColor,
    borderRadius:
      props.borderRadius ??
      defaultNotificationTypeConfig?.borderRadius ??
      (defaultGlobalConfig?.borderRadius || themeBase.borderRadius.default),
    borderColor:
      props.borderColor ??
      defaultNotificationTypeConfig?.borderColor ??
      (defaultGlobalConfig?.borderColor || chooseAccentColor(notificationType)),
    borderWidth:
      props.borderWidth ??
      defaultNotificationTypeConfig?.borderWidth ??
      defaultGlobalConfig?.borderWidth ??
      1,
    multiline:
      props.multiline ?? defaultNotificationTypeConfig?.multiline ?? defaultGlobalConfig?.multiline,
    defaultIconType:
      props.defaultIconType ??
      defaultNotificationTypeConfig?.defaultIconType ??
      defaultGlobalConfig?.defaultIconType,
    leftIconSource:
      customIconSource ??
      chooseDefaultIcon(
        notificationType,
        darkMode,
        props.defaultIconType ??
          defaultNotificationTypeConfig?.defaultIconType ??
          defaultGlobalConfig?.defaultIconType
      ),
  }
}
