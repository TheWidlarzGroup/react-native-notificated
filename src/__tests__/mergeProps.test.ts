import { mergeProps } from '../defaultConfig/mergeProps'

test('get proper props for component', () => {
  const props = {
    description: 'test description props',
    descriptionColor: 'rgb(0, 0, 50)',
    borderRadius: 5,
    borderType: 'no-border',
  }

  const globalConfig = {
    description: 'test description global',
    descriptionSize: 5,
    descriptionColor: 'rgb(0, 100, 0)',
    borderRadius: 5,
  }

  const notificationTypeConfig = {
    description: 'test description',
    titleSize: 23,
    titleColor: '#fff',
    descriptionSize: 10,
    descriptionColor: 'rgb(0, 0, 0)',
    borderRadius: 5,
  }

  expect(mergeProps(props, 'error', true, globalConfig, notificationTypeConfig)).toMatchSnapshot()
})
