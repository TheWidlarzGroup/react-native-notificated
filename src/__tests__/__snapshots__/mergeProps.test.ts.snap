// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`get proper props for component 1`] = `
Object {
  "accentColor": "#FC6060",
  "bgColor": undefined,
  "borderRadius": 5,
  "borderType": "border",
  "borderWidth": 1,
  "defaultIconType": undefined,
  "description": "test description props",
  "descriptionColor": "rgb(0, 0, 0)",
  "descriptionSize": 10,
  "leftIconSource": Object {
    "testUri": "../../../src/assets/images/error-color.png",
  },
  "multiline": undefined,
  "onPress": undefined,
  "theme": "dark",
  "title": "",
  "titleColor": "#fff",
  "titleSize": 23,
}
`;
